doc/man/man3/listen.3.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man3/listen.3.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH LISTEN 3 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Library Calls"
.\"
.\"
.SH NAME
.B listen
\- listen for socket connections and limit the queue of incoming connections
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/socket.h>
.HP
.BI "int listen(int " socket ", int " backlog );
.\"
.\"
.SH ARGUMENTS
.PP
The function takes the following arguments:
.TP
.I socket
Specifies the file descriptor associated with the socket upon which to listen.
.TP
.I backlog
Specifies the maximum number of queued (backlogged) connection indications that
should be permitted to wait.
.\"
.\"
.SH DESCRIPTION
.PP
The
.BR listen ()
function marks a connection-mode socket, specified by the
.I socket
argument, as accepting connections.
.PP
The
.I backlog
argument provides a hint to the implementation which the
implementation will use to limit the number of outstanding connections in the
socket's listen queue. Normally, a larger
.I backlog
argument value will result
in a larger or equal length of the listen queue.
.PP
The implementation may include incomplete connections in its listen queue. The
limits on the number of incomplete connections and completed connections
queued may be different.
.PP
The implementation may have an upper limit on the length of the listen
queue\(emeither global or per accepting socket.  If
.I backlog
exceeds this limit, the
length of the listen queue is set to the limit.
.PP
If
.BR listen ()
is called with a
.I backlog
argument value that is less than 0, the
function behaves as if it had been called with a
.I backlog
argument value of 0.
.PP
A
.I backlog
argument of 0 may allow the
.I socket
to accept connections, in which
case the length of the listen queue may be set to an implementation-dependent
minimum value.
.PP
The
.I socket
in use may require the process to have appropriate privileges to
use the
.BR listen ()
function.
.\"
.\"
.SH CAVEATS
.PP
This system call is implemented as a
.B SIOCSOCKSYS
IOCTL.  See
.BR socksys (4)
for more information.
.\"
.\"
.SH "RETURN VALUE"
.PP
Upon successful completions,
.BR listen ()
returns 0. Otherwise, -1 is returned and
errno is set to indicate the error.
.\"
.\"
.SH ERRORS
.PP
The
.BR listen ()
function will fail if:
.TP
.RB [ EBADF ]
The
.I socket
argument is not a valid file descriptor.
.TP
.RB [ EDESTADDRREQ ]
The
.I socket
is not bound to a local address, and the protocol does not support
listening on an unbound
.IR socket .
.TP
.RB [ EINVAL ]
The
.I socket
is already connected.
.TP
.RB [ ENOTSOCK ]
The
.I socket
argument does not refer to a socket.
.TP
.RB [ EOPNOTSUPP ]
The
.I socket
protocol does not support
.BR listen ()
.PP
The
.BR listen ()
function may fail if:
.TP
.RB [ EACCES ]
The calling process does not have the appropriate privileges.
.TP
.RB [ EINVAL ]
The
.I socket
has been shut down.
.TP
.RB [ ENOBUFS ]
Insufficient resources are available in the system to complete the call.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR accept (3),
.BR connect (3),
.BR socket (3),
.BR <sys/socket.h> (5).
.\"
.\"
.SH COMPATIBILITY
.PP
This function and sockets implementation is compatible to
.IR "SVR 4.2" ,
.[
svr42
.]
and implementations based on
.IR "SVR 4.2" ,
including
.IR Solaris \(rg,
.[
soldoc
.]
with the following portability considerations:
.IP \(em \w'\(em\(em'u
Where 4BSD differs from POSIX, this function follows POSIX.
.[
susv3
.]
.PP
See
.BR sockets (3)
and
.BR socklib (3)
for more compatibility information.
.\"
.\"
.SH CONFORMANCE
.PP
The Open Group XNS Issue 5.2.
.[
xns98
.]
.\"
.\"
.SH HISTORY
.PP
Most sockets functions first appeared in
.IR "4BSD" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


isode/others/listen/listen.c
/*****************************************************************************

 @(#) $RCSfile$ $Name$($Revision$) $Date$

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2007  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date$ by $Author$

 -----------------------------------------------------------------------------

 $Log$
 *****************************************************************************/

#ident "@(#) $RCSfile$ $Name$($Revision$) $Date$"

static char const ident[] = "$RCSfile$ $Name$($Revision$) $Date$";

/* listen.c -- responder for listen demo */

#ifndef	lint
static char *rcsid =
    "Header: /xtel/isode/isode/others/listen/RCS/listen.c,v 9.0 1992/06/16 12:42:15 isode Rel";
#endif

/* 
 * Header: /xtel/isode/isode/others/listen/RCS/listen.c,v 9.0 1992/06/16 12:42:15 isode Rel
 *
 *
 * Log: listen.c,v
 * Revision 9.0  1992/06/16  12:42:15  isode
 * Release 8.0
 *
 */

/*
 *				  NOTICE
 *
 *    Acquisition, use, and distribution of this module and related
 *    materials are subject to the restrictions of a license agreement.
 *    Consult the Preface in the User's Manual for the full terms of
 *    this agreement.
 *
 */

#include <stdio.h>
#include "listen.h"

/*  */

static char *mycontext = "isode listen demo";

/*  */

/* ARGSUSED */

main(argc, argv, envp)
	int argc;
	char **argv, **envp;
{
	int secs, vecp;
	char *vec[4];
	struct TSAPdisconnect tds;
	struct TSAPdisconnect *td = &tds;
	struct SSAPstart sss;
	register struct SSAPstart *ss = &sss;
	struct SSAPdata sxs;
	register struct SSAPdata *sx = &sxs;
	struct SSAPindication sis;
	register struct SSAPindication *si = &sis;
	register struct SSAPabort *sa = &si->si_abort;
	AEI aei;
	struct PSAPaddr *pa;
	struct sblk incoming;
	SB sbi = &incoming;

	reportailor(argv[0]);

	if (argc > 2)
		adios(NULLCP, "usage: %s [secs]", argv[0]);

	secs = (argc == 2) ? atoi(argv[1]) : 15;	/* Wait 15s after release by default */

	if ((aei = str2aei(PLocalHostName(), mycontext)) == NULLAEI)
		adios(NULLCP, "%s-%s: unknown application-entity", PLocalHostName(), mycontext);

	if ((pa = aei2addr(aei)) == NULLPA)
		adios(NULLCP, "address translation failed");

	if (TNetListen(&pa->pa_addr.sa_addr, td) == NOTOK)
		if (td->td_cc > 0)
			adios(NULLCP, "TNetListen: [%s] %*.*s", TErrString(td->td_reason),
			      td->td_cc, td->td_cc, td->td_data);
		else
			adios(NULLCP, "TNetListen: [%s]", TErrString(td->td_reason));

	/* now wait for incoming call */
	for (;;) {
		if (TNetAccept(&vecp, vec, 0, NULLFD, NULLFD, NULLFD, NOTOK, td)
		    == NOTOK) {
			if (td->td_cc > 0)
				adios(NULLCP, "TNetAccept: [%s] %*.*s", TErrString(td->td_reason),
				      td->td_cc, td->td_cc, td->td_data);
			else
				adios(NULLCP, "TNetAccept: [%s]", TErrString(td->td_reason));
		}

		if (vecp > 0)
			break;
	}

	if (SInit(vecp, vec, ss, si) == NOTOK)
		adios(NULLCP, "S-CONNECT.INDICATION: %s", SErrString(sa->sa_reason));
	advise(LLOG_NOTICE, NULLCP,
	       "S-CONNECT.INDICATION: <%d, %s, %s, %s, %s, %ld, %d>",
	       ss->ss_sd, sprintref(&ss->ss_connect),
	       saddr2str(&ss->ss_calling), saddr2str(&ss->ss_called),
	       sprintb(ss->ss_requirements, RMASK), ss->ss_isn, ss->ss_ssdusize);

	/* stop listening, we have what we want */
	(void) TNetClose(NULLTA, td);

	bzero((char *) sbi, sizeof *sbi);
	sbi->sb_sd = ss->ss_sd;
	sbi->sb_connect = ss->ss_connect;	/* struct copy */
	sbi->sb_requirements = ss->ss_requirements & SR_BASUBSET;
	sbi->sb_settings = ss->ss_settings;
#define dotoken(requires,shift,bit,type) \
{ \
    if (sbi -> sb_requirements & requires) \
	switch (sbi -> sb_settings & (ST_MASK << shift)) { \
	    case ST_CALL_VALUE << shift: \
		sbi -> sb_settings &= ~(ST_MASK << shift); \
		sbi -> sb_settings |= ST_INIT_VALUE << shift; \
		break; \
 \
	    case ST_INIT_VALUE: \
		break; \
 \
	    case ST_RESP_VALUE: \
		sbi -> sb_owned |= bit; \
		break; \
 \
	    default: \
		adios (NULLCP, "%s token: reserved", type); \
		break; \
	} \
}
	dotokens();
#undef	dotoken
	sbi->sb_ssn = sbi->sb_isn = ss->ss_isn;

	SSFREE(ss);

	if (SConnResponse(sbi->sb_sd, &sbi->sb_connect, NULLSA, SC_ACCEPT,
			  sbi->sb_requirements, sbi->sb_settings, sbi->sb_isn,
			  NULLCP, 0, si) == NOTOK)
		adios(NULLCP, "S-CONNECT.RESPONSE: %s", SErrString(sa->sa_reason));

/* do work here */

	switch (SReadRequest(sbi->sb_sd, sx, secs, si)) {
	case NOTOK:
		adios(NULLCP, "S-READ.REQUEST: %s", SErrString(sa->sa_reason));

	case OK:
		adios(NULLCP, "not expecting DATA indication 0x%x", sx->sx_type);

	case DONE:
		if (si->si_type != SI_FINISH)
			adios(NULLCP, "not expecting indication 0x%x", si->si_type);
		if (SRelResponse(sbi->sb_sd, SC_ACCEPT, NULLCP, 0, si) == NOTOK)
			adios(NULLCP, "S-RELEASE.RESPONSE: %s", SErrString(sa->sa_reason));
		break;
	}

	exit(0);
	return 0;
}
